httpcore5/src/test/java/org/apache/hc/core5/http/TestContentType.java

```diff
@@ -183,4 +183,132 @@ void testWithParams() throws Exception {
+
+        // Testing multipart/form-data should not include charset in toString
+        contentType = ContentType.create("multipart/form-data", StandardCharsets.UTF_8, true);
+        Assertions.assertEquals("multipart/form-data", contentType.toString());
+        Assertions.assertNotNull(contentType.getCharset());
+    }
+
+
+    @Test
+    void testCharsetForOtherMediaTypes() {
+        // Testing application/json should include charset
+        ContentType contentType = ContentType.create("application/json", StandardCharsets.UTF_8, false);
+        Assertions.assertEquals("application/json; charset=UTF-8", contentType.toString());
+        Assertions.assertEquals(StandardCharsets.UTF_8, contentType.getCharset());
+
+        // Testing text/html should include charset
+        contentType = ContentType.create("text/html", StandardCharsets.UTF_8, false);
+        Assertions.assertEquals("text/html; charset=UTF-8", contentType.toString());
+        Assertions.assertEquals(StandardCharsets.UTF_8, contentType.getCharset());
+    }
+
+    @Test
+    void testNoCharsetForBinaryMediaTypes() throws Exception {
+        // Test for application/octet-stream
+        ContentType contentType = ContentType.create("application/octet-stream", null, true);
+        Assertions.assertEquals("application/octet-stream", contentType.toString());
+        Assertions.assertNull(contentType.getCharset());
+
+        // Test for image/jpeg
+        contentType = ContentType.create("image/jpeg", null, true);
+        Assertions.assertEquals("image/jpeg", contentType.toString());
+        Assertions.assertNull(contentType.getCharset());
+    }
+
+    @Test
+    void testFormUrlEncodedWithoutCharset() throws Exception {
+        // Test for application/x-www-form-urlencoded with percent-encoding
+        final ContentType contentType = ContentType.create("application/x-www-form-urlencoded", null, true);
+        Assertions.assertEquals("application/x-www-form-urlencoded", contentType.toString());
+        Assertions.assertNull(contentType.getCharset());
+
+        // Test body encoding example with percent-encoding
+        final String encodedBody = "echotext=TEST%F6TEST";
+        // Simulate HTTP redirect where charset shouldn't change
+        final String redirectedBody = "echotext=TEST%F6TEST";
+        Assertions.assertEquals(encodedBody, redirectedBody);  // Ensure body stays the same after redirect
+    }
+
+    @Test
+    void testImplicitCharsetForApplicationJson() throws Exception {
+        // ContentType for application/json with implicitCharset = true
+        final ContentType contentType = ContentType.create("application/json", StandardCharsets.UTF_8, true);
+
+        // Check that the charset is not added to the toString() output
+        Assertions.assertEquals("application/json", contentType.toString());
+        // Check that the charset is still stored
+        Assertions.assertEquals(StandardCharsets.UTF_8, contentType.getCharset());
+    }
```

Comment:
What about `foo/bar`, null, false`? With a null charset the last param should not matter.